server/server_test.go

```diff
@@ -770,6 +774,138 @@ func TestLogSourceIPs(t *testing.T) {
+
+	helloPathParamRouteTmpl := "/hello/{pathParam}"
+	// unnamed routes will be labeled with their registered path template
+	server.HTTP.HandleFunc(helloPathParamRouteTmpl, writeSpanToBodyHandleFunc)
+
+	// extracted route names or path templates are converted to a Prometheus-compatible label value
+	expectedHelloRouteLabel := middleware.MakeLabelValue(helloRouteName)
+	expectedHelloPathParamRouteLabel := middleware.MakeLabelValue(helloPathParamRouteTmpl)
+
+	go func() {
+		require.NoError(t, server.Run())
+	}()
+
+	helloRouteURL := "http://127.0.0.1/hello"
+	expectedTagsHelloRoute := map[string]interface{}{
+		"component":       "net/http",
+		"span.kind":       "server",
+		"http.url":        helloRouteURL,
+		"http.method":     "GET",
+		"http.route":      expectedHelloRouteLabel,
+		"http.user_agent": "",
+	}
+	helloPathParamRouteURL := "http://127.0.0.1/hello/world"
+	expectedTagsHelloPathParamRoute := map[string]interface{}{
+		"component":       "net/http",
+		"span.kind":       "server",
+		"http.url":        helloPathParamRouteURL,
+		"http.method":     "GET",
+		"http.route":      expectedHelloPathParamRouteLabel,
+		"http.user_agent": "",
+	}
+
+	target := server.GRPCListenAddr()
+	conn, err := grpc.Dial(
+		target.String(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(4*1024*1024)),
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+	client := httpgrpc.NewHTTPClient(conn)
+
+	// emulateHTTPGRPCPRoxy mimics the usage of the Server type as a load balancing proxy,
+	// wrapping http requests into gRPC requests to utilize gRPC load balancing features
+	emulateHTTPGRPCPRoxy := func(
+		client httpgrpc.HTTPClient, req *http.Request,
+	) (*httpgrpc.HTTPResponse, error) {
+		require.NoError(t, err)
+		req.RequestURI = req.URL.String()
+		grpcReq, err := httpgrpcServer.HTTPRequest(req)
+		require.NoError(t, err)
+		return client.Handle(req.Context(), grpcReq)
+	}
+
+	{
+		req, err := http.NewRequest(
+			"GET", helloRouteURL, bytes.NewReader([]byte{}),
+		)
+		require.NoError(t, err)
+		resp, err := emulateHTTPGRPCPRoxy(client, req)
+		require.NoError(t, err)
+		var bodyJSON tracingSpanJSON
+		err = json.NewDecoder(bytes.NewReader(resp.Body)).Decode(&bodyJSON)
+		require.NoError(t, err)
+		require.Equal(t, "HTTP GET - "+expectedHelloRouteLabel, bodyJSON.OperationName)
+		require.Equal(t, expectedTagsHelloRoute, bodyJSON.Tags)
+	}
+	{
+		req, err := http.NewRequest(
+			"GET", helloPathParamRouteURL, bytes.NewReader([]byte{}),
+		)
+		require.NoError(t, err)
+		resp, err := emulateHTTPGRPCPRoxy(client, req)
+		require.NoError(t, err)
+		var bodyJSON tracingSpanJSON
+		err = json.NewDecoder(bytes.NewReader(resp.Body)).Decode(&bodyJSON)
+		require.NoError(t, err)
+		require.Equal(t, "HTTP GET - "+expectedHelloPathParamRouteLabel, bodyJSON.OperationName)
+		require.Equal(t, expectedTagsHelloPathParamRoute, bodyJSON.Tags)
+	}
```

Comment:
Might be clearer to wrap these two test cases in a call to `t.Run(...)` each.

These two cases could also be set up as data-driven test cases, which would help reduce the amount of duplication for each case.